Alerting: Receiver API complete core implementation

JacobsonMT opened the pull request:
**What is this feature?**

Includes the finished core implementation of the new receivers API, backing receiver service, and authentication service.

Includes:
- Receiver model improvements: centralization of common receiver patterns (encryption, decryption, merging settings, ...) on the model itself. Removed `SecureFields` on the model layer as it's meaning was vague and error-prone. These changes fixed some edge cases in redaction that can be seen in test changes.
- Lots of new tests to cover new and existing receiver functionality.
- Optimistic concurrency.
- UID-scoped read/write/create/delete actions on receivers (and a migration to move the legacy implicit global actions to explicit global scoped actions).
- Validation that receivers are not used by rules or routes before deletion.
- Renaming of references to receivers in routes and rule notification settings in receiver service (though this is not yet supported through the API)
- Error message improvements (though there is still room for improvement).

**Why do we need this feature?**

Migrate to resource specific APIs for notification CRUD.

**Who is this feature for?**
Front-end developers

**Special notes for your reviewer:**

Most commits (other than early refactorings) are best reviewed separately. Can split commits into separate PRs if desired.

Not included in this PR:
- Extra metadata on the receiver response:
  - Permissions.
  - Information about count / which rules/routes use the receivers.
  - Type information.

JacobsonMT committed:
Replace global authz abstraction with one compatible with uid scope

JacobsonMT committed:
Replace GettableApiReceiver with models.Receiver in receiver_svc

JacobsonMT committed:
GrafanaIntegrationConfig -> models.Integration

JacobsonMT committed:
Implement Create/Update methods

JacobsonMT committed:
Add optimistic concurrency to receiver API

JacobsonMT committed:
Add scope to ReceiversRead & ReceiversReadSecrets

migrates existing permissions to include implicit global scope

JacobsonMT committed:
Add receiver create, update, delete actions

JacobsonMT committed:
Check if receiver is used by rules before delete

JacobsonMT committed:
On receiver name change update in routes and notification settings

JacobsonMT committed:
Improve errors

JacobsonMT committed:
Linting

yuri-tceretian commented on pkg/services/ngalert/notifier/receiver_svc.go:
```diff
@@ -271,26 +279,162 @@ func (rs *ReceiverService) DeleteReceiver(ctx context.Context, uid string, orgID
+	})
+}
+
+func (rs *ReceiverService) CreateReceiver(ctx context.Context, r *models.Receiver, orgID int64, user identity.Requester) (*models.Receiver, error) {
```
I think we should perform struct validation before any other activities, i.e. validate against alerting integrations.

yuri-tceretian commented on pkg/services/ngalert/notifier/receiver_svc.go:
```diff
@@ -58,61 +84,60 @@ func NewReceiverService(
+		encryptionService:      encryptionService,
+		xact:                   xact,
+		log:                    log,
+		validator:              validation.ValidateProvenanceRelaxed,
```
let's rename to provenanceValidator.

yuri-tceretian commented on pkg/registry/apis/alerting/notifications/receiver/legacy_storage.go:
```diff
@@ -179,16 +190,16 @@ func (s *legacyStorage) Update(ctx context.Context,
 	}
 
-	if p.ObjectMeta.Name != getUID(model) {
+	if p.ObjectMeta.Name != model.GetUID() {
```
The domain service supports renaming. Let's remove this blocker?

yuri-tceretian commented on pkg/services/ngalert/notifier/receiver_svc.go:
```diff
@@ -271,26 +279,162 @@ func (rs *ReceiverService) DeleteReceiver(ctx context.Context, uid string, orgID
+	err = rs.xact.InTransaction(ctx, func(ctx context.Context) error {
+		// If the name of the receiver changed, we must update references to it in both routes and notification settings.
+		if existing.Name != r.Name {
+			affected, err := rs.ruleNotificationsStore.RenameReceiverInNotificationSettings(ctx, orgID, existing.Name, r.Name)
```
I think this should be a little bit more complex. we should check provenance status compatibility. For example, if we rename a receiver via UI but rules are provisioned, this call should be rejected.

yuri-tceretian commented on pkg/services/ngalert/notifier/legacy_storage/compat.go:
```diff
@@ -15,3 +22,42 @@ func UidToName(uid string) (string, error) {
+
+func IntegrationToPostableGrafanaReceiver(integration *models.Integration) (*apimodels.PostableGrafanaReceiver, error) {
+	postable := &apimodels.PostableGrafanaReceiver{
+		UID:                   integration.UID,
```
Here or somewhere else upstream we should generate new GUID. I could not find any code in the new service that does this. 

Also, we need to make sure that this GUID is unique across the Alertmanager configuration

yuri-tceretian commented on pkg/services/ngalert/notifier/receiver_svc.go:
```diff
@@ -271,26 +279,162 @@ func (rs *ReceiverService) DeleteReceiver(ctx context.Context, uid string, orgID
+}
+
+func (rs *ReceiverService) UpdateReceiver(ctx context.Context, r *models.Receiver, storedSecureFields map[string][]string, orgID int64, user identity.Requester) (*models.Receiver, error) {
+	if err := rs.authz.AuthorizeUpdate(ctx, user, r); err != nil {
```
I think here is the problem: we authorize access to receiver with UID(name1) and rename name1 -> name2, which effectively changes the UID of the resulting receiver. 

authorization should detect whether user has access to old and new UID.

yuri-tceretian reviewed (commented):
I think I am half way through the review. Please see my comments so far.

These are just my notes for later (follow-up PRs) :
- We need to add support for selectors
- Provisioning service is intact
- Add permissions to the creator

JacobsonMT committed:
Include read permissions are requirements for create/update/delete

JacobsonMT commented on pkg/registry/apis/alerting/notifications/receiver/authorize.go:
```diff
@@ -20,19 +20,24 @@ func Authorize(ctx context.Context, ac accesscontrol.AccessControl, attr authori
 		action = accesscontrol.EvalAny(
-			accesscontrol.EvalPermission(accesscontrol.ActionAlertingNotificationsWrite), // TODO: Add alert.notifications.receivers:delete permission
+			accesscontrol.EvalPermission(accesscontrol.ActionAlertingNotificationsWrite),
+			accesscontrol.EvalPermission(accesscontrol.ActionAlertingReceiversDelete),
```
It would be simple to add the `ReceiverAccess` service here, what's not clear to me is what kind of responsibility it should have. For example:
- It is responsible for completely determining auth:
  - This is easy enough when the request is for a single resource with a UID, but what about those without a clear name. For example: does `create` have a value in`attr.GetName()`, `list` response filtering would need to happen after the list is retrieved.
- It is responsible for just preconditon checks:
  - This should be straightforward but the benefits aren't clear to me as it's redundant with the legacy storage and not sufficient for the unified storage.

yuri-tceretian replied:
I think the authz process should be multi-tier, sooner we reject an unauthorized request, better, the authorization logic in domain service also makes sense to prevent unauthorized access via dependencies.
I think create action does not need UID scope because it does not seem to make sense because it's awkward to be able to pre-authorize creation of a receiver with a specific name. List operation can still use `readRedactedReceiversPreConditionsEval`

JacobsonMT committed:
Alias ngalert/models to ngmodels to differentiate from v0alpha1 model

JacobsonMT committed:
Ensure integration UIDs are valid, unique, and generated if empty

JacobsonMT committed:
Validate integration settings on create/update

JacobsonMT committed:
Leverage UidToName to GetReceiver instead of GetReceivers

JacobsonMT committed:
Remove some unnecessary uses of simplejson

JacobsonMT committed:
alerting.notifications.receiver -> alerting.notifications.receivers

JacobsonMT committed:
validator -> provenanceValidator

JacobsonMT committed:
Only validate the modified receiver

stops existing invalid receivers from preventing modification of a valid
receiver.

yuri-tceretian commented on pkg/services/ngalert/accesscontrol/receivers.go:
```diff
@@ -68,16 +72,108 @@ var (
+	// Asserts create access to all receivers.
+	createAllReceiversEval = ac.EvalAny(
+		ac.EvalPermission(ac.ActionAlertingNotificationsWrite),
+		ac.EvalPermission(ac.ActionAlertingReceiversCreate, ScopeReceiversAll),
```
I think create operation does not need scopes.

JacobsonMT committed:
Improve error in Integration.Encrypt

JacobsonMT committed:
Remove scope from alert.notifications.receivers:create

JacobsonMT committed:
Add todos for receiver renaming

JacobsonMT committed:
Use receiverAC precondition checks in k8s api

JacobsonMT committed:
Merge branch 'main' into jacobsonmt/receiver-api-all

# Conflicts:
#	pkg/services/ngalert/models/testing.go

JacobsonMT committed:
Linting

JacobsonMT committed:
Optional optimistic concurrency for delete

JacobsonMT committed:
Improve error message on authorizer.DecisionDeny

JacobsonMT committed:
make update-workspace

JacobsonMT committed:
Merge branch 'main' into jacobsonmt/receiver-api-all

yuri-tceretian commented on pkg/services/ngalert/notifier/receiver_svc.go:
```diff
@@ -253,9 +253,12 @@ func (rs *ReceiverService) DeleteReceiver(ctx context.Context, uid string, calle
+		err = rs.checkOptimisticConcurrency(existing, version)
+		if err != nil {
+			return err
+		}
```
It would be great if we could do something like this, and log the situation when optimistic concurrency was skipped

https://github.com/grafana/grafana/blob/135f6571a9b8a5c471282b575e5fa5f670c3622e/pkg/services/ngalert/provisioning/mute_timings.go#L391-L404

JacobsonMT committed:
More specific auth checks in k8s authorize.go

JacobsonMT committed:
Add debug log when delete optimistic concurrency is skipped

yuri-tceretian reviewed (approved):
LGTM! Epic job!

JacobsonMT committed:
Keep error for non-forbidden errutil errors

JacobsonMT committed:
Merge branch 'main' into jacobsonmt/receiver-api-all
